compiler/luci/logex/src/CircleNodeSummaryBuilder.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __LUCI_LOGEX_CIRCLE_NODE_SUMMARY_BUILDER__
+#define __LUCI_LOGEX_CIRCLE_NODE_SUMMARY_BUILDER__
+
+#include "CircleNodeSummaryBuilderHelper.h"
+
+#include <luci/IR/CircleDialect.h>
+#include <luci/IR/CircleNode.h>
+#include <locop/NodeSummary.h>
+#include <locop/SymbolTable.h>
+
+#include <memory>
+#include <sstream>
+#include <vector>
+
+namespace luci
+{
+
+class CircleNodeSummaryBuilder
+{
+public:
+  bool build(const loco::Node *node, const locop::SymbolTable *tbl, locop::NodeSummary &s)
+  {
+    if (node->dialect() != luci::CircleDialect::get())
+      return false;
+
+    auto ptr_to_str = [](const void *ptr) {
+      std::stringstream ss;
+      ss << ptr;
+      return ss.str();
+    };
+
+    auto circle_node = loco::must_cast<const luci::CircleNode *>(node);
+    if (const auto builder = create_builder(circle_node))
```

Comment:
I'm not sure I understand this correctly but this seems to create `CircleNodeSummaryBuilder` object every time a node has to print summary of the graph.
Could you please check how may times this is called for a Inception V3 model (any version) with `LUCI_LOG=100` set ?

Reply:
319 (nodes in IV3) * 12 (printing function called) = 3828 creation and I think this is not a big deal :D

---

FYI, current code is creating 4272 times T.T
It's because
- `SummaryBuilderLet<SB::GRP>(_tbl).build(node, s))` creates `SummaryBuilderLet<SB::GRP>`
- According to the type of node, `SummaryBuilderLet<SB::GRP>` can be created with maximum 9 trials.

https://github.com/Samsung/ONE/blob/4c9870d3b781fe1f422e082d4bdee3b1a1638d0e/compiler/luci/logex/src/FormattedGraph.cpp#L2241-L2256

Reply:
I'd like to understand if current code do object allocation.
New `create_builder` will create an object `3828 ` times while running(circle2circle I guess).

> current code is creating 4272 times

To compare with `create_builder` which code is using allocation?

Reply:
`SummaryBuilderLet<SB::GRP>(_tbl).build(node, s))` creates `SummaryBuilderLet<SB::GRP>`.
AFAIK, unless the class isn't Singleton(or static), it creates new class everytime.
So following two codes are same.

```cpp
   do
   {
     if (SummaryBuilderLet<SB::GRP>(_tbl).build(node, s))
       return true;
   } while (false)
```

```cpp
   do
   {
     auto builder = SummaryBuilderLet<SB::GRP>(_tbl);
     if (builder.build(node, s))
       return true;
   } while (false)
```

Am I misunderstand something...?

Reply:
> Am I misunderstand something...?

I'm also not sure... One thing to check is either
`SummaryBuilderLet<SB::GRP>(_tbl).build(node, s))` or `auto builder = SummaryBuilderLet<SB::GRP>(_tbl);`, where the allocation happens.
`make_unique` will use heap, but not sure about `SummaryBuilderLet<SB::GRP>(_tbl);`
If this is created in the heap, it will be the same allocation. Number is decreased so this is good. If this is created in the stack, memory management will be less heavy.

Anyway, as this is used when the LOG is enabled for debugging, concern is not critical.

BTW, is it possible not to use make_unique/make_shared?

Reply:
> make_unique will use heap, but not sure about SummaryBuilderLet<SB::GRP>(_tbl);

It would be created on stack :)

```cpp
auto on_heap = new SummaryBuilderLet<SB::GRP>(_tbl); // SummaryBuilderLet<SB::GRP>*
auto on_stack = SummaryBuilderLet<SB::GRP>(_tbl); // SummaryBuilderLet<SB::GRP>
```

> BTW, is it possible not to use make_unique/make_shared?

Unfortunately, not using pointer is impossible T.T
One way to avoid using `make_unique` is just using raw pointer.
However, if we use raw pointer, memory leak can be occurred :(

Reply:
OK, let's go as-is. When/if this becomes issue, let's revise this later.